lib/puppet/type/x509_cert.rb

```diff
@@ -64,6 +64,26 @@
     end
   end
 
+  newparam(:is_server, :boolean => true) do
```

Comment:
Maybe name this `is_ca` if I understood the goal correctly?

Reply:
I believe, there is a difference between being a CA and the certificate type server (vs. client).
My openssl.cnf (RHEL 7.5) has the following section where the certType can be set.
```
[ usr_cert ]
# These extensions are added when 'ca' signs a request.
# This goes against PKIX guidelines but some CAs do it and some software
# requires this to avoid interpreting an end user certificate as a CA.
basicConstraints=CA:FALSE
# Here are some examples of the usage of nsCertType. If it is omitted
# the certificate can be used for anything *except* object signing.
# This is OK for an SSL server.
# nsCertType                    = server
# For an object signing certificate this would be used.
# nsCertType = objsign
# For normal client use this is typical
# nsCertType = client, email
[...]
```
When generating a CA, the following section is used:
```
[ v3_ca ]
# Extensions for a typical CA
# PKIX recommendation.
subjectKeyIdentifier=hash
authorityKeyIdentifier=keyid:always,issuer
# This is what PKIX recommends but some broken software chokes on critical
# extensions.
#basicConstraints = critical,CA:true
# So we do this instead.
basicConstraints = CA:true
# Key usage: this is typical for a CA certificate. However since it will
# prevent it being used as an test self-signed certificate it is best
# left out by default.
# keyUsage = cRLSign, keyCertSign
# Some might want this also
# nsCertType = sslCA, emailCA
# Include email address in subject alt name: another PKIX recommendation
# subjectAltName=email:copy
# Copy issuer details
# issuerAltName=issuer:copy
# DER hex encoding of an extension: beware experts only!
# obj=DER:02:03
# Where 'obj' is a standard or added object
# You can even override a supported extension:
# basicConstraints= critical, DER:30:03:01:01:FF
```
So, if @WetHippie is referring to generating a CA certificate, using is_ca would make sense. As for the nsCertType being an array, I would prefer to have an Enum to pass, instead of a boolean for just one setting. And as the comments in the openssl file hints, it requires a CA for signing as these fields are added upon signing. I don't know if the field will be in a self-signed certificate.

This change might also require the correct settings in the config template as well.

Reply:
Right, it seems indeed that it should be an enum instead of a boolean in that case.